### PR TITLE
Adding upper limit on number of z-ranges generated

### PR DESCRIPTION
**Pull request by elahrvivaz**
- Range calculation is changed to breadth-first
- Ranges for multiple search intervals can be generated at once
- Z2/Z3 code consolidated where possible

Note: I initially extracted ZRange into a trait and kept the z2 and z3 specific flavors, but for some reason that caused a large slow-down in unit tests in GeoMesa. Instead, I moved the operations in the Z2/Z3 objects.

### CODE REVIEW

**Comment by jahhulbert-ccri** [comment]
interesting...i wonder why the trait would be so slow.

File: zorder/src/main/scala/org/locationtech/sfcurve/zorder/ZRange.scala [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+  // contains in index space (e.g. the long value)
+  def contains(r: ZRange): Boolean =  contains(r.min) && contains(r.max)
+
+  // overlaps in index space (e.g. the long value)
+  def overlaps(r: ZRange): Boolean = contains(r.min) || contains(r.max)
+}
+
+object ZRange {
+  def apply(min: Z2, max: Z2)(implicit d: DummyImplicit): ZRange = ZRange(min.z, max.z)
```
Comment (jnh5y) [inline zorder/src/main/scala/org/locationtech/sfcurve/zorder/ZRange.scala:37]: Ok, did these implicits remain from some intermediate work?

File: zorder/src/main/scala/org/locationtech/sfcurve/zorder/ZRange.scala [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+ * Represents a rectangle in defined by min and max as two opposing points
+ * 
+ * @param min: lower-left point 
+ * @param max: upper-right point
+ */
+case class ZRange(min: Long, max: Long) {
+
+  require(min <= max, s"Range bounds must be ordered, but $min > $max")
+
+  def mid: Long = (max + min)  >>> 1 // overflow safe mean
```
Comment (jnh5y) [inline zorder/src/main/scala/org/locationtech/sfcurve/zorder/ZRange.scala:22]: Does this work if max + min > overflow?

**Comment by jnh5y** [comment]
@lossyrob @echeipesh do you gents have any time to take a look at this PR?

Since it renames things quite a bit, I'm tempted to bump a minor version in respect of semver.  I'm also happy to hear an argument that the library is young enough that staying on the 0.1.x series is sensible.

**Comment by jnh5y** [comment]
@lossyrob, as a quick legal concern, we might want to use this PR as a time to figure out how to handle copyright headers for this project.

@elahrvivaz is using the 'creator-of-the-file' copyrights.  I'm fine with that or having multiple authors/entities listed in the header.  Thoughts?

**Comment by lossyrob** [comment]
@jnh5y The 'creator of the file' works for us.

**Comment by lossyrob** [comment]
I would also be fine for bumping the minor version. +1

**Comment by jnh5y** [comment]
@lossyrob thanks!  If you aren't fussed, I'll merge this tomorrow and cut a 0.2.0 release.

**Comment by lossyrob** [comment]
Do it to it!

File: zorder/src/main/scala/org/locationtech/sfcurve/zorder/ZRange.scala (hunk 1 above)
Comment (elahrvivaz) [inline zorder/src/main/scala/org/locationtech/sfcurve/zorder/ZRange.scala:37]: no, b/c Z3 and Z2 are both value classes, the runtime signature for both applies is `apply(Long, Long)`. In order to differentiate the two methods we use the dummy implicits.

File: zorder/src/main/scala/org/locationtech/sfcurve/zorder/ZRange.scala (hunk 2 above)
Comment (elahrvivaz) [inline zorder/src/main/scala/org/locationtech/sfcurve/zorder/ZRange.scala:22]: yep! that's the whole idea. Chris originally found this I believe.

**Comment by elahrvivaz** [comment]
@jnh5y @lossyrob I didn't realize the original files were copyright azavea - we should probably preserve both copyrights.

**Comment by jnh5y** [comment]
Committed as 3018fbc89f2a613cb2b97edc6a91ef07f7102021 after bumping to 0.2.0-SNAPSHOT.

Also, tagged and cut 0.2.0.

**Pull request closed without merging**